lessons/module-1/js-pure-functions.md

```diff
@@ -0,0 +1,133 @@
+
+<section class="answer">
+### Impure vs Pure Function Example
+
+**Impure Function:**
+```js
+var userEmail = "jeremiah@turing.io";
+
+function confirmEmail() {
+  console.log(`Your email address is currently ${userEmail}`);
+}
+
+confirmEmail();  // "Your email address is currently jeremiah@turing.io"
+
+userEmail = "jeremiah@turing.edu";
+confirmEmail(); // "Your email address is currently jeremiah@turing.edu"
+```
+
+**Pure Function:**
+```js
+function confirmEmail(userEmail) {
+  console.log(`Your email address is currently ${userEmail}`);
+}
+
+confirmEmail("jeremiah@turing.io") // "Your email address is currently jeremiah@turing.io"
+
+confirmEmail("jeremiah@turing.edu") // "Your email address is currently jeremiah@turing.edu"
+```
+</section>
+
+<section class="call-to-action">
+### On Your Own
+
+Based on the above examples and defintions, answer the following:
+
+1. What makes the first function impure?  What is the side effect?
+2. In the second example, what makes the function pure?  If you call the function with the same argument, will you always get the same result?
+3. What are the pros of making pure functions?  Are there any downsides?
+4. **Spicy:** 🌶️ Is it possible to make all functions pure?  What is an example where it would be impossible to make a function pure? 🌶️
+</section>
+
+<section class="note">
+### Notes To Remember
+By keeping much of the logic of an application built on **pure functions**, it can make your code more predictable and safer.  Ultimately this will also allow your application to be easier to test and debug.  Following this structure can allow us to follow a more functional paradigm with your JS which can be significantly easier to use than other paradigms such as OOP (Object Oriented Programming).
+
+**Pure functions** will always have a `return` statement as well as arguments/parameters.  Remember a pure function returns the same output if given the same input.  If there is no `return` statement, this is likely because it is updating something else in the "outside" world.
+
+Not all functions are going to be pure.  Functions that must mutate data outside of the function's scope, update the DOM, or rely on some outside source are impure.  This is *okay*.  The goal is to make as much of the program as pure as possible and save the functions that require side effects for last!
+</section>
+
+## Getting More Practice
+
+Let's look at a few more examples.  In breakout groups, you will look at the next 3 exercises and determine if the function is pure or impure.  If the function is impure, refactor it so that the function is pure!
+
+<section class="call-to-action">
+### In Your Notebook
+
+Answer the following questions for the next 3 exercises:
+1. Is the function pure?  Why or why not?
+2. If the function is impure, what would need to be changed to make it pure?
```

Comment:
Would love to add a question about testing pure functions.

"What do you think will be easier to test, a pure function or an impure function? Why?"

Reply:
Yes!!!  Good callout, I'll add that in!